File: harness/go/harness.go
// Trace harness: captures median nanosecond costs for the six crime
// operations and four rescue operations used in the honestcode.software
// demo engine.
//
// Runs N iterations, discards warmup, reports median per operation.
// Uses batched operations (100 iterations per measurement) to overcome
// macOS timer granularity (~42ns).
//
// Usage:
//
//	go run harness.go
package main

import (
	"encoding/json"
	"fmt"
	"os"
	"runtime"
	"sort"
	"sync"
	"time"
)

const (
	warmup = 200
	runs   = 1000
	batch  = 100 // operations per measurement to overcome timer granularity
)

// ═══════════════════════════════════════════════
// CRIME SCENE: mutable struct with singleton pattern
// ═══════════════════════════════════════════════

type CouponRegistry struct{}

var (
	couponOnce     sync.Once
	couponInstance *CouponRegistry
)

func GetCouponRegistry() *CouponRegistry {
	couponOnce.Do(func() {
		couponInstance = &CouponRegistry{}
	})
	return couponInstance
}

func (r *CouponRegistry) Lookup(code string) float64 {
	if code == "SAVE10" {
		return 0.10
	}
	return 0
}

type TaxService struct{}

var (
	taxOnce     sync.Once
	taxInstance *TaxService
)

func GetTaxService() *TaxService {
	taxOnce.Do(func() {
		taxInstance = &TaxService{}
	})
	return taxInstance
}

func (s *TaxService) Calculate(region string, taxable float64) float64 {
	return taxable * 0.08
}

type Item struct {
	Name  string
	Price float64
}

type Order struct {
	Items      []Item
	Total      float64
	Discount   float64
	Tax        float64
	CouponCode string
	UpdatedAt  time.Time
}

// crimeRun measures each operation over a batch and returns per-op nanoseconds.
func crimeRun() [6]int64 {
	items := []Item{
		{"Widget", 29.99},
		{"Gadget", 39.99},
		{"Doohickey", 19.99},
	}

	// call: method dispatch (append items)
	start := time.Now()
	for b := 0; b < batch; b++ {
		order := &Order{}
		order.Items = append(order.Items, items...)
		_ = order
	}
	callNs := time.Since(start).Nanoseconds() / batch

	// field: mutable field write
	start = time.Now()
	for b := 0; b < batch; b++ {
		order := &Order{}
		order.CouponCode = "SAVE10"
		order.Discount = 0
		_ = order
	}
	fieldNs := time.Since(start).Nanoseconds() / batch

	// calc: computation (loop over collection)
	order := &Order{Items: items}
	start = time.Now()
	for b := 0; b < batch; b++ {
		total := 0.0
		for _, item := range order.Items {
			total += item.Price
		}
		order.Total = total
	}
	calcNs := time.Since(start).Nanoseconds() / batch

	// single: singleton lookup (sync.Once)
	start = time.Now()
	for b := 0; b < batch; b++ {
		couponOnce = sync.Once{}
		couponInstance = nil
		taxOnce = sync.Once{}
		taxInstance = nil
		_ = GetCouponRegistry()
		_ = GetTaxService()
	}
	singleNs := time.Since(start).Nanoseconds() / batch / 2 // per lookup

	// cache: cache check (method call with string comparison)
	registry := GetCouponRegistry()
	start = time.Now()
	for b := 0; b < batch; b++ {
		rate := registry.Lookup("SAVE10")
		order.Discount = order.Total * rate
	}
	cacheNs := time.Since(start).Nanoseconds() / batch

	// time: timestamp capture
	start = time.Now()
	for b := 0; b < batch; b++ {
		order.UpdatedAt = time.Now()
	}
	timeNs := time.Since(start).Nanoseconds() / batch

	// prevent dead code elimination
	taxService := GetTaxService()
	taxable := order.Total - order.Discount
	order.Tax = taxService.Calculate("NY", taxable)
	_ = order.Tax

	return [6]int64{callNs, fieldNs, calcNs, singleNs, cacheNs, timeNs}
}

// ═══════════════════════════════════════════════
// RESCUE: pure functions, flat data
// ═══════════════════════════════════════════════

type OrderResult struct {
	Total    float64
	Tax      float64
	Subtotal float64
}

type FinalResult struct {
	Total      float64
	Tax        float64
	Subtotal   float64
	CouponCode string
	Discount   float64
	GrandTotal float64
}

func calculateOrder(items []Item, region string, taxRates map[string]float64) OrderResult {
	total := 0.0
	for _, item := range items {
		total += item.Price
	}
	rate, ok := taxRates[region]
	if !ok {
		rate = 0
	}
	tax := total * rate
	return OrderResult{Total: total, Tax: tax, Subtotal: total + tax}
}

func applyCoupon(order OrderResult, code string, coupons map[string]float64) FinalResult {
	rate, ok := coupons[code]
	if !ok {
		rate = 0
	}
	discount := order.Total * rate
	return FinalResult{
		Total:      order.Total,
		Tax:        order.Tax,
		Subtotal:   order.Subtotal,
		CouponCode: code,
		Discount:   discount,
		GrandTotal: order.Subtotal - discount,
	}
}

func rescueRun() [4]int64 {
	items := []Item{
		{"Widget", 29.99},
		{"Gadget", 39.99},
		{"Doohickey", 19.99},
	}
	taxRates := map[string]float64{"NY": 0.08, "CA": 0.0725}
	coupons := map[string]float64{"SAVE10": 0.10}

	// call: function call overhead (empty call pair)
	start := time.Now()
	for b := 0; b < batch; b++ {
		_ = time.Now()
	}
	callNs := time.Since(start).Nanoseconds() / batch

	// arg: argument passing
	start = time.Now()
	for b := 0; b < batch; b++ {
		region := "NY"
		_ = region
	}
	argNs := time.Since(start).Nanoseconds() / batch

	// calc: pure computation
	start = time.Now()
	var result OrderResult
	for b := 0; b < batch; b++ {
		result = calculateOrder(items, "NY", taxRates)
	}
	calcNs := time.Since(start).Nanoseconds() / batch

	// ret: return value construction
	start = time.Now()
	var final FinalResult
	for b := 0; b < batch; b++ {
		final = applyCoupon(result, "SAVE10", coupons)
	}
	retNs := time.Since(start).Nanoseconds() / batch

	// prevent dead code elimination
	if final.GrandTotal < 0 {
		panic("impossible")
	}

	return [4]int64{callNs, argNs, calcNs, retNs}
}

// ═══════════════════════════════════════════════
// HARNESS: warmup, collect, median
// ═══════════════════════════════════════════════

func medianInt64(values []int64) int64 {
	sort.Slice(values, func(i, j int) bool { return values[i] < values[j] })
	return values[len(values)/2]
}

type output struct {
	Language string           `json:"language"`
	OS       string           `json:"os"`
	Runtime  string           `json:"runtime"`
	Runs     int              `json:"runs"`
	Crime    map[string]int64 `json:"crime"`
	Rescue   map[string]int64 `json:"rescue"`
}

func main() {
	// Warmup
	for i := 0; i < warmup; i++ {
		crimeRun()
		rescueRun()
	}

	// Collect
	crimeResults := make([][6]int64, runs)
	rescueResults := make([][4]int64, runs)
	for i := 0; i < runs; i++ {
		crimeResults[i] = crimeRun()
		rescueResults[i] = rescueRun()
	}

	// Compute medians
	crimeOps := []string{"call", "field", "calc", "single", "cache", "time"}
	rescueOps := []string{"call", "arg", "calc", "ret"}

	crime := make(map[string]int64)
	for op := 0; op < len(crimeOps); op++ {
		col := make([]int64, runs)
		for i := 0; i < runs; i++ {
			col[i] = crimeResults[i][op]
		}
		crime[crimeOps[op]] = medianInt64(col)
	}

	rescue := make(map[string]int64)
	for op := 0; op < len(rescueOps); op++ {
		col := make([]int64, runs)
		for i := 0; i < runs; i++ {
			col[i] = rescueResults[i][op]
		}
		rescue[rescueOps[op]] = medianInt64(col)
	}

	out := output{
		Language: "go",
		OS:       fmt.Sprintf("%s %s", runtime.GOOS, runtime.GOARCH),
		Runtime:  runtime.Version(),
		Runs:     runs,
		Crime:    crime,
		Rescue:   rescue,
	}

	enc := json.NewEncoder(os.Stdout)
	enc.SetIndent("", "  ")
	enc.Encode(out)
}

File: harness/python/harness.py
"""
Trace harness: captures median nanosecond costs for the six crime
operations and four rescue operations used in the honestcode.software
demo engine.

Runs N iterations, discards warmup, reports median per operation.

Usage:
    python harness.py
"""
import json
import platform
import statistics
import sys
import time


WARMUP = 200
RUNS = 1000


# ═══════════════════════════════════════════════
# CRIME SCENE: mutable Order class with singletons
# ═══════════════════════════════════════════════

class CouponRegistry:
    _instance = None

    @classmethod
    def get_instance(cls):
        if cls._instance is None:
            cls._instance = cls()
        return cls._instance

    def lookup(self, code):
        if code == "SAVE10":
            return 0.10
        return 0


class TaxService:
    _instance = None

    @classmethod
    def get_instance(cls):
        if cls._instance is None:
            cls._instance = cls()
        return cls._instance

    def calculate(self, region, taxable):
        return taxable * 0.08


class Order:
    def __init__(self):
        self.items = []
        self.total = 0.0
        self.discount = 0.0
        self.tax = 0.0
        self.coupon_code = None
        self.updated_at = None


def crime_run():
    """Measure individual crime operations. Returns [call, field, calc, single, cache, time]."""
    order = Order()
    items = [
        {"name": "Widget", "price": 29.99},
        {"name": "Gadget", "price": 39.99},
        {"name": "Doohickey", "price": 19.99},
    ]
    ns = time.perf_counter_ns

    # call: method dispatch (simulate addItem call overhead)
    t0 = ns()
    order.items.extend(items)
    call_ns = ns() - t0

    # field: mutable field write
    t0 = ns()
    order.coupon_code = "SAVE10"
    order.discount = 0.0
    field_ns = ns() - t0

    # calc: computation (sum over collection via attribute access)
    t0 = ns()
    order.total = sum(item["price"] for item in order.items)
    calc_ns = ns() - t0

    # single: singleton lookup
    t0 = ns()
    registry = CouponRegistry.get_instance()
    tax_service = TaxService.get_instance()
    single_ns = (ns() - t0) // 2  # per lookup

    # cache: cache check (hash-based lookup)
    t0 = ns()
    rate = registry.lookup(order.coupon_code)
    order.discount = order.total * rate
    cache_ns = ns() - t0

    # time: timestamp capture
    t0 = ns()
    order.updated_at = time.time_ns()
    time_ns = ns() - t0

    # calc tax to prevent dead code elimination
    taxable = order.total - order.discount
    order.tax = tax_service.calculate("NY", taxable)

    return [call_ns, field_ns, calc_ns, single_ns, cache_ns, time_ns]


# ═══════════════════════════════════════════════
# RESCUE: pure functions, flat data
# ═══════════════════════════════════════════════

def calculate_order(items, region, tax_rates):
    total = sum(item["price"] for item in items)
    tax = total * tax_rates.get(region, 0)
    return total, tax, total + tax


def apply_coupon(total, subtotal, code, coupons):
    rate = coupons.get(code, 0)
    discount = total * rate
    return code, discount, subtotal - discount


def rescue_run():
    """Measure individual rescue operations. Returns [call, arg, calc, ret]."""
    items = [
        {"name": "Widget", "price": 29.99},
        {"name": "Gadget", "price": 39.99},
        {"name": "Doohickey", "price": 19.99},
    ]
    tax_rates = {"NY": 0.08, "CA": 0.0725}
    coupons = {"SAVE10": 0.10}
    ns = time.perf_counter_ns

    # call: function call overhead
    t0 = ns()
    t1 = ns()
    call_ns = t1 - t0

    # arg: argument passing
    t0 = ns()
    region = "NY"
    arg_ns = ns() - t0

    # calc: pure computation (returns tuples, not dicts)
    t0 = ns()
    total, tax, subtotal = calculate_order(items, region, tax_rates)
    calc_ns = ns() - t0

    # ret: apply coupon (returns tuple, no allocation beyond that)
    t0 = ns()
    code, discount, grand_total = apply_coupon(total, subtotal, "SAVE10", coupons)
    ret_ns = ns() - t0

    # prevent dead code elimination
    assert grand_total > 0

    return [call_ns, arg_ns, calc_ns, ret_ns]


# ═══════════════════════════════════════════════
# HARNESS: warmup, collect, median
# ═══════════════════════════════════════════════

def main():
    # Warmup
    for _ in range(WARMUP):
        crime_run()
        rescue_run()

    # Collect
    crime_results = [crime_run() for _ in range(RUNS)]
    rescue_results = [rescue_run() for _ in range(RUNS)]

    # Compute medians
    crime_ops = ["call", "field", "calc", "single", "cache", "time"]
    rescue_ops = ["call", "arg", "calc", "ret"]

    crime_medians = {}
    for i, op in enumerate(crime_ops):
        col = [r[i] for r in crime_results]
        crime_medians[op] = statistics.median(col)

    rescue_medians = {}
    for i, op in enumerate(rescue_ops):
        col = [r[i] for r in rescue_results]
        rescue_medians[op] = statistics.median(col)

    output = {
        "language": "python",
        "os": f"{platform.system()} {platform.release()}",
        "runtime": f"CPython {platform.python_version()}",
        "runs": RUNS,
        "crime": crime_medians,
        "rescue": rescue_medians,
    }

    print(json.dumps(output, indent=2))


if __name__ == "__main__":
    main()
